src/main/java/org/ethereum/beacon/discovery/storage/KBucket.java

```diff
@@ -0,0 +1,162 @@
+  }
+
+  public Optional<NodeRecord> getPendingNode() {
+    return pendingNode.map(BucketEntry::getNode);
+  }
+
+  public void offer(final NodeRecord node) {
+    performMaintenance();
+    getEntry(node)
+        .ifPresentOrElse(
+            existing -> updateExistingRecord(existing, node), () -> offerNewNode(node));
+  }
+
+  private void updateExistingRecord(final BucketEntry existing, final NodeRecord newRecord) {
+    if (existing.getNode().getSeq().compareTo(newRecord.getSeq()) >= 0) {
+      // New record isn't actually newer, do nothing.
+      return;
+    }
+    nodes.remove(existing);
+    final BucketEntry newEntry = new BucketEntry(livenessChecker, newRecord);
+    nodes.add(newEntry);
+    newEntry.checkLiveness(clock.millis());
+  }
+
+  private void offerNewNode(final NodeRecord node) {
+    if (isFull()) {
+      getLastNode().checkLiveness(clock.millis());
+      if (pendingNode.isEmpty()) {
+        livenessChecker.checkLiveness(node);
+      }
+    } else {
+      final BucketEntry newEntry = new BucketEntry(livenessChecker, node);
+      nodes.add(newEntry);
+      newEntry.checkLiveness(clock.millis());
+    }
+  }
+
+  public void onNodeContacted(final NodeRecord node) {
+    getEntry(node)
+        .ifPresentOrElse(
+            existingEntry -> {
+              // Move to the start of the bucket
+              nodes.remove(existingEntry);
+              nodes.add(0, existingEntry.withLastConfirmedTime(clock.millis()));
+              performMaintenance();
+            },
+            () -> {
+              if (pendingNode.isPresent()
+                  && pendingNode.get().getNodeId().equals(node.getNodeId())) {
+                // Update pending node
+                pendingNode = Optional.of(pendingNode.get().withLastConfirmedTime(clock.millis()));
+              }
+              performMaintenance();
+              if (isFull()) {
+                if (pendingNode.isEmpty()) {
+                  pendingNode = Optional.of(new BucketEntry(livenessChecker, node, clock.millis()));
+                }
+              } else {
+                nodes.add(0, new BucketEntry(livenessChecker, node, clock.millis()));
+              }
+            });
+  }
+
+  /**
+   * Performs any pending maintenance on the bucket.
+   *
+   * <p>If the pending node has not been pinged recently, schedule a ping for it.
+   *
+   * <p>If the pending node has not responded to the last ping within a reasonable time, remove it.
+   *
+   * <p>If the last node in the bucket has not been pinged recently, schedule a ping for it.
+   *
+   * <p>If the last node in the bucket has not responded to the last liveness check within a
+   * reasonable time:
+   *
+   * <p>a. remove it from the bucket.
+   *
+   * <p>b. if there is a pending node, insert it into the bucket (at appropriate position based on
+   * when it was last confirmed as live)
+   */
+  public void performMaintenance() {
+    if (nodes.isEmpty()) {
+      return;
+    }
```

Comment:
we might want to allow somehow getting into a state where nodes is empty, and we have a pending...

in that case we don't want to return here, we would just want to move past the nodes maintenance and onto the pending logic...

It would be a little sad skipping if we have a pending node that's valid and we just never insert it because the current list size is 0...

Reply:
I'm pretty sure it's never possible to have an empty nodes list and a pending node (and it would definitely be a bug if it happened) but yes, seems reasonable to always check the pending nodes here just in case.